File: python/tests/person_id_corpus.py
"""The person-id grammar corpus: 74 rows, every expectation a measured Go verdict.

Not derived from documentation and not from this implementation. A probe linked
against the real ``go/pkg/types.FlexibleInt64`` and the real
``normalizeEmbeddedPeopleJSON`` (``go/pkg/basecamp/normalize.go:40``, the
``coercePersonID`` rule) produced every row, so a disagreement here is a
disagreement with Go rather than with a reading of it.

One table, two sites. The rule is ``strconv.ParseInt(s, 10, 64)`` and its three
outcomes, and both of the SDK's person-id sites read the SAME row differently
only in what they DO with the outcome:

===========  =============================================  ==========================
``kind``     the pre-decode normalizer must                 the flexible reader must
===========  =============================================  ==========================
``"value"``  write the int, no ``system_label``             return the int
``"label"``  write id ``0`` and ``system_label`` = the raw  return 0
``"refuse"`` leave the string untouched                     fail the read
===========  =============================================  ==========================

Rows that exist to discriminate, so do not drop them as redundant: ``"+7"`` and
``"+007"`` (the sign a ``^-?\\d+$`` regex refuses); ``"007"``, ``"010"``,
``"0009223372036854775807"`` (leading zeros -- ``"010"`` is TEN, and Ruby's
``Integer()`` said eight); the Unicode digit rows (Python's ``int()`` parses
every one of them); the whitespace rows (``int()`` strips ASCII whitespace); the
underscore rows (PEP 515, which ``int()`` accepts and base-10 ``ParseInt`` never
does); ``"18446744073709551615x"`` against ``"18446744073709551616x"`` -- one
digit apart, OPPOSITE refusals, because Go checks the magnitude inside the scan
against ``uint64`` and the first disqualifying byte wins; and
``"9007199254740992"`` / ``"9007199254740993"``, past JS's safe-integer range
but real ``int64`` ids.
"""

from __future__ import annotations

#: ``(raw, kind, value)``. ``value`` is Go's own number for a ``"value"`` row
#: and ``None`` for the other two, which carry no number.
PERSON_ID_CORPUS: list[tuple[str, str, int | None]] = [
    # --- plain, signed and zero-padded: ParseInt's whole accepting grammar ---
    ("7", "value", 7),
    ("0", "value", 0),
    ("-0", "value", 0),
    ("+0", "value", 0),
    ("+7", "value", 7),
    ("-7", "value", -7),
    ("007", "value", 7),
    ("+007", "value", 7),
    ("-007", "value", -7),
    ("0009223372036854775807", "value", 9223372036854775807),
    ("0000000000000000000000009", "value", 9),
    ("010", "value", 10),  # TEN. Base 10 has no octal prefix.
    ("00", "value", 0),
    ("0000", "value", 0),
    ("-00", "value", 0),
    # --- an empty digit run, with or without a sign ---
    ("", "label", None),
    ("+", "label", None),
    ("-", "label", None),
    # --- whitespace: Go trims NONE of it ---
    (" ", "label", None),
    (" 7", "label", None),
    ("7 ", "label", None),
    (" 7 ", "label", None),
    ("\n7", "label", None),
    ("7\n", "label", None),
    ("\t7", "label", None),
    ("7\t", "label", None),
    # --- PEP 515 underscores: a base-0 feature in Go, never a base-10 one ---
    ("1_0", "label", None),
    ("1_2", "label", None),
    # --- base prefixes: also base-0 features ---
    ("0x10", "label", None),
    ("0b11", "label", None),
    ("0o17", "label", None),
    ("0X1F", "label", None),
    # --- junk around or instead of the digits ---
    ("7x", "label", None),
    ("x7", "label", None),
    ("12.0", "label", None),
    ("1e3", "label", None),
    ("12,3", "label", None),
    # --- the system actors this whole rule exists to name ---
    ("basecamp", "label", None),
    ("campfire", "label", None),
    ("LocalPerson", "label", None),
    # --- Unicode decimal digits: ASCII 0-9 only, so none of these is a digit ---
    ("１２３", "label", None),  # fullwidth 123
    ("７", "label", None),  # fullwidth 7
    ("٠١٢", "label", None),  # Arabic-Indic 012
    ("٠", "label", None),  # Arabic-Indic 0
    ("৭", "label", None),  # Bengali 7
    ("۷", "label", None),  # Extended Arabic-Indic 7
    ("৭7", "label", None),  # Bengali 7 then ASCII 7
    ("7৭", "label", None),  # ASCII 7 then Bengali 7
    # --- the int64 boundary, from both sides ---
    ("9223372036854775806", "value", 9223372036854775806),
    ("9223372036854775807", "value", 9223372036854775807),
    ("9223372036854775808", "refuse", None),
    ("9223372036854775809", "refuse", None),
    ("-9223372036854775807", "value", -9223372036854775807),
    ("-9223372036854775808", "value", -9223372036854775808),  # the one asymmetric row
    ("-9223372036854775809", "refuse", None),
    ("+9223372036854775807", "value", 9223372036854775807),
    ("+9223372036854775808", "refuse", None),
    ("0000000000000000000009223372036854775807", "value", 9223372036854775807),
    # --- past int64 but still inside uint64: ParseInt's own bound refuses them ---
    ("18446744073709551614", "refuse", None),
    ("18446744073709551615", "refuse", None),
    ("18446744073709551616", "refuse", None),
    ("99999999999999999999999", "refuse", None),
    ("00000000000000000000018446744073709551616", "refuse", None),
    # --- scan order: which refusal fires depends on WHERE the bad byte sits ---
    # One digit apart, opposite answers. "...615x" gets to the 'x' with the
    # accumulator at exactly uint64 max, so it is a SYNTAX error and reads as
    # the system actor; "...616x" overflows uint64 on the last digit, before
    # the scan ever looks at the 'x', so it is a RANGE error and fails.
    # Testing the string for well-formedness first gets this pair backwards --
    # in the accepting direction, which turns an unreadable id into "basecamp".
    ("18446744073709551615x", "label", None),
    ("18446744073709551616x", "refuse", None),
    ("1844674407370955161x", "label", None),
    ("-18446744073709551615x", "label", None),
    ("-18446744073709551616x", "refuse", None),
    ("99999999999999999999999x", "refuse", None),
    # --- beyond JS's safe integers, but ordinary int64 person ids ---
    ("9007199254740991", "value", 9007199254740991),
    ("9007199254740992", "value", 9007199254740992),
    ("9007199254740993", "value", 9007199254740993),
    ("90071992547409931", "value", 90071992547409931),
    ("-9007199254740993", "value", -9007199254740993),
]

File: typescript/tests/helpers/person-id-corpus.ts
/**
 * The measured person-id corpus: 74 strings, and the verdict Go reaches on each.
 *
 * Every row is a MEASUREMENT, not a reading of the docs: a probe linked against
 * the real `go/pkg/types.FlexibleInt64` and the real `normalizeEmbeddedPeopleJSON`
 * produced the third column. `strconv.ParseInt(s, 10, 64)` has exactly three
 * outcomes and the two sites that read a string person id in this SDK — the
 * pre-decode normalizer in `src/services/base.ts` and the read-side comparison in
 * `src/services/mentions.ts` — each map all three:
 *
 * | Go says | the normalizer writes                  | the reader returns |
 * |---------|----------------------------------------|--------------------|
 * | value n | the number n, no `system_label`        | n                  |
 * | SYNTAX  | id `0` + `system_label` = the raw text | 0                  |
 * | RANGE   | nothing — the string stays put         | `undefined`        |
 *
 * `value` carries a `bigint` on purpose. Go's answer is an `int64`, and eleven
 * of these rows name one a JS `number` cannot hold; writing them here as
 * `number` literals would round the reference itself and the table would then
 * agree with the port by construction. The tests derive what each site may do
 * with an unrepresentable value from `fitsNumber` below, so every residual
 * divergence is visible as a residual rather than hidden in a rounded constant.
 *
 * Rows that exist to discriminate, and must not be dropped as redundant:
 * `"+7"`/`"+007"` (the sign a `^-?\d+$` regex refuses and ParseInt takes);
 * `"007"`, `"010"`, `"0009223372036854775807"` (leading zeros carry no meaning —
 * `"010"` is TEN); the Unicode digits; the whitespace forms Go trims none of;
 * the `_` separators base 10 does not take; `"18446744073709551615x"` against
 * `"18446744073709551616x"` (one digit apart, opposite refusals — the scan-order
 * pair); and `"9007199254740992"`/`"9007199254740993"`, real int64 ids past
 * JS's safe-integer range.
 */
export type CorpusRow = {
  readonly id: string;
  /** What `strconv.ParseInt(id, 10, 64)` does with it. */
  readonly go: "value" | "syntax" | "range";
  /** The int64 Go read, for `go: "value"` rows only. */
  readonly value?: bigint;
};

/** Whether a JS `number` can hold Go's answer without rounding it. */
export function fitsNumber(value: bigint): boolean {
  return value >= BigInt(Number.MIN_SAFE_INTEGER) && value <= BigInt(Number.MAX_SAFE_INTEGER);
}

const value = (id: string, v: bigint): CorpusRow => ({ id, go: "value", value: v });
const syntax = (id: string): CorpusRow => ({ id, go: "syntax" });
const range = (id: string): CorpusRow => ({ id, go: "range" });

export const PERSON_ID_CORPUS: readonly CorpusRow[] = [
  value("7", 7n),
  value("0", 0n),
  value("-0", 0n),
  value("+0", 0n),
  value("+7", 7n),
  value("-7", -7n),
  value("007", 7n),
  value("+007", 7n),
  value("-007", -7n),
  value("0009223372036854775807", 9223372036854775807n),
  value("0000000000000000000000009", 9n),
  syntax(""),
  syntax(" "),
  syntax("+"),
  syntax("-"),
  syntax(" 7"),
  syntax("7 "),
  syntax(" 7 "),
  syntax("\n7"),
  syntax("7\n"),
  syntax("\t7"),
  syntax("7\t"),
  syntax("1_0"),
  syntax("1_2"),
  syntax("0x10"),
  syntax("0b11"),
  syntax("0o17"),
  value("010", 10n),
  syntax("0X1F"),
  syntax("7x"),
  syntax("x7"),
  syntax("12.0"),
  syntax("1e3"),
  syntax("12,3"),
  syntax("basecamp"),
  syntax("campfire"),
  syntax("LocalPerson"),
  syntax("１２３"),
  syntax("７"),
  syntax("٠١٢"),
  syntax("৭"),
  syntax("۷"),
  value("9223372036854775806", 9223372036854775806n),
  value("9223372036854775807", 9223372036854775807n),
  range("9223372036854775808"),
  range("9223372036854775809"),
  value("-9223372036854775807", -9223372036854775807n),
  value("-9223372036854775808", -9223372036854775808n),
  range("-9223372036854775809"),
  range("18446744073709551614"),
  range("18446744073709551615"),
  range("18446744073709551616"),
  syntax("18446744073709551615x"),
  range("18446744073709551616x"),
  syntax("1844674407370955161x"),
  syntax("-18446744073709551615x"),
  range("-18446744073709551616x"),
  range("99999999999999999999999"),
  range("99999999999999999999999x"),
  range("00000000000000000000018446744073709551616"),
  value("0000000000000000000009223372036854775807", 9223372036854775807n),
  value("9007199254740991", 9007199254740991n),
  value("9007199254740992", 9007199254740992n),
  value("9007199254740993", 9007199254740993n),
  value("90071992547409931", 90071992547409931n),
  value("-9007199254740993", -9007199254740993n),
  value("+9223372036854775807", 9223372036854775807n),
  range("+9223372036854775808"),
  value("00", 0n),
  value("0000", 0n),
  value("-00", 0n),
  syntax("٠"),
  syntax("৭7"),
  syntax("7৭"),
];
